### ISSUE TITLE
So laggy

### ISSUE THREAD

**Issue by o02140o** [Qalculate/qalculate-gtk#250]
macOS Big Sur Version 11.2.3
MacBook Air (M1, 2020)
installed with homebrew (iTerm+rosetta2)
"brew install qalculate-gtk"
after finish i opened it using terminal (i've try with rosetta and without)
the result is so laggy
there's delay about 500ms
how to fix it?

**Comment by hanna-kn** [comment]
Do you mean that there is a noticeable delay before the result is displayed (after Enter or equals button is pressed)?
Or before the main window appears (500 ms seems insignificant in this case)?
Unfortunately I do not have access to any macOS system, and I am therefore not able to reproduce and find the cause of this issue myself.

**Comment by o02140o** [comment]
it took about 10 second to open (ik because it run using rosetta), But i think that's ok for me.
The main problem is every i type (anything (number/word)) it got lagg/delay about 500ms(0,5 second) per number on my MBA m1, (i've try on windows laptop it worked well, but not on my MBA m1) 
*(mba=macbook air) 
*m1=apple silicon processor (architecture: ARM)

**Comment by hanna-kn** [comment]
I see no logical reason for Qalculate to be _that_ slow because of Rosetta.

The performance should improve if you disable completion and expression status.

Have you tried any other software that uses Gtk3 and/or was installed using Homebrew?

**Comment by demosn01** [comment]
Same issue on macOS 11 2019 macbook pro. Extreamly laggy and unusable. After I believe version 3.16 but can be sure. I am not quite sure how to reinstall the older versions. Installed using homebrew

**Comment by veloxo** [comment]
I'm experiencing this on Intel MacOS Big Sur 11.2.3, just installed via Homebrew.

The lag is only present when the keypad is visible. The UI performs normally after collapsing the keypad or switching to minimal-window mode.

**Comment by demosn01** [comment]
> 
> The lag is only present when the keypad is visible. The UI performs normally after collapsing the keypad or switching to minimal-window mode.
> 

Yes, I just tried @veloxo suggestion that the lag occurs when the keyboard is **visible**. **Hiding the keyboard** makes the app not laggy. Thank you @veloxo :) Hopefully a dev can fix it in the future

**Comment by o02140o** [comment]
(qalculate-gtk:3043): Gtk-CRITICAL **: 17:51:57.550: gtk_tree_view_scroll_to_cell: assertion 'tree_view->priv->tree != NULL' failed

With Keypad: https://drive.google.com/file/d/1XcsIaKRRtzJMqgVOesYxbEYMJHh7v9KR/view?usp=sharing

Without Keypad: https://drive.google.com/file/d/1TcHo7dRLLaIaJbnFO8_lFI53nehmp4qw/view?usp=sharing

it worked flawlessly if i close the keypad.

btw why dont u make qalculate on appstore? it would be great, and you can earn some money

**Comment by hanna-kn** [comment]
Does it make any difference (at all) if the right or left side of the keypad is hidden, or if the history is visible instead?

**Comment by demosn01** [comment]
> Does it make any difference (at all) if the right or left side of the keypad is hidden, or if the history is visible instead?

Hi @hanna-kn, Yes it does make a difference. Hiding one side of the keyboard reveals a "cumulatory" lag effect.

What this means is **hiding one side of the keyboard** makes the lag a bit more tolerable (approximately **half the lag time**). Hiding the other side has the exact same effect. **When both keyboards are open, the lag is the greatest**.

When the **history is open**. It has the **exact same effect as hiding one side of the of keyboard** (ie. there is still a lag but maybe half as much)

I am assuming this means it has something to do with the UI? Would you be able to shed some light on what may be going on? Thank you for any info!

**Comment by veloxo** [comment]
Similar results here, but the History lag is barely perceptible. I only notice it when holding down backspace, or when clicking the menu bar the flyout appears instantly but the blue highlighting lags a moment.

Keyboard open = **100 lag**
Hiding one keyboard = **~50 lag**
Histroy open = **5 lag**
Conversion open = **zero lag**

**Comment by hanna-kn** [comment]
My theory is that the entire window (and its contents) is updated each time the expression entry is updated. You can check if this is true by running qalculate-gtk with `--debug-gtk=updates`. This might be unreasonably slow because of issues in the Cairo and/or GTK libraries on macOS.

I discovered that the whole window is updated when the expression entry changes from empty to non-empty, and vice versa, (when the icon in the upper right corner is hidden/shown), in versions >= 3.16. This does, however, not provide a likely explanation for this issue.

**Comment by demosn01** [comment]
> My theory is that the entire window (and its contents) is updated each time the expression entry is updated. You can check if this is true by running qalculate-gtk with `--debug-gtk=updates`. This might be unreasonably slow because of issues in the Cairo and/or GTK libraries on macOS.
> 
> I discovered that the whole window is updated when the expression entry changes from empty to non-empty, and vice versa, (when the icon in the upper right corner is hidden/shown), in versions >= 3.16. This does, however, not provide a likely explanation for this issue.

Hi @hanna-kn 
Thank you very much for the info.
I can't seem to activate the qalculate-gtk with --debug-gtk=updates. I did a google search, tried --help but I cant get it to work.
### This is what I ran in the terminal:
`qalculate-gtk --debug-gtk=updates` which gives the output message: **Unknown option --debug-gtk=updates**.
I also tried `qalculate-gtk --enable-debug=yes` but same thing.

The debug option is not listed in `qalculate-gtk --help` which I think is strange?. I also found people having the same sortof problem but with different software. https://bugs.launchpad.net/ubuntu/+source/evolution/+bug/1718613

Sorry @hanna-kn. I tried to get it to work but not there yet.

**Comment by hanna-kn** [comment]
I made a mistake when writing the command line option. The correct command is `qalculate-gtk --gtk-debug=updates`.

**Comment by demosn01** [comment]
> I made a mistake when writing the command line option. The correct command is `qalculate-gtk --gtk-debug=updates` (see https://developer.gnome.org/gtk3/stable/gtk-running.html).

Thats okay, thank you. I found that url and tried that too. It still doesn't work for `qalculate-gtk --gtk-debug=updates`. Same error of: **Unknown option --gtk-debug=updates**

**Comment by hanna-kn** [comment]
The command line option works for me on Windows and various Linux distribution.
`GTK_DEBUG=updates qalculate-gtk` should achieve the same result (but it is possible that it does not do anything if the command line option is not present).

**Comment by demosn01** [comment]
Sorry for late reply @hanna-kn. This command works on mac `GTK_DEBUG=updates qalculate-gtk` but appears **G_ENABLED_DEBUG** is not enabled for mac.

output is this:
(qalculate-gtk:3891): Gtk-WARNING **: 02:59:05.062: GTK_DEBUG set but ignored because gtk isn't built with G_ENABLE_DEBUG

The commands you gave upbove work on linux just fine btw

**Comment by m-zheng** [comment]
This issue still exists when qalculate-gtk and libqalculate are updated to 3.20.1. But can be temporarily solved by @veloxo method.
